### PR TITLE
Adds support for replacing base type references with their derived types

### PR DESCRIPTION
**Pull request by irvinesunday**
Related to: https://github.com/microsoftgraph/microsoft-graph-explorer-api/issues/196

Proposes:
- Adding functionality to replace base type `$refs` in `reponses` and/or `requestBody` payloads with their derived types `$refs` in the `oneOf` property of their schemas.
- Adding config. settings for toggling on/off the feature for either the `responses` payload or `requestBody` payloads. This is toggled to `false` by default.
- Adding an internal static class, `Helper.cs`, that has an internal static method that is reusable by operation methods that have the above functionalities toggled on to get the derived types `$refs` in the `oneOf` property of their `schemas`. 

**Notes**:
This feature will be used alongside the `x-ms-discriminator-value` extension (for derived types) and the `discriminator` object annotations (for base types) by AutoRest through the OpenAPIService (of the [DevXAPI](https://graphexplorerapi.azurewebsites.net/openapi/operations)) that slices the OpenAPI doc returned by this lib., based on operation ids or request urls. The OpenAPIService needs to have the derived types described in the schema so that AutoRest would be able to generate polymorphic types that return appropriate types based on the `@odata.type` for consumption by the [Microsoft Graph PowerShell SDK](https://github.com/microsoftgraph/msgraph-sdk-powershell).

### CODE REVIEW

**Comment by xuzhg** [comment]
@irvinesunday can we have some test cases to cover these changes? and meanwhile, would you please rebase them to master branch

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs [hunk 1]
```diff
@@ -471,6 +472,44 @@ public void CreatePropertySchemaForNullableEnumPropertyReturnSchema(OpenApiSpecV
...
         #endregion
 
+        #region BaseTypeToDerivedTypesSchema
+
+        [Fact]
+        public void GetDerivedTypesReferenceSchemaReturnsDerivedTypesReferencesInSchemaIfExist()
+        {
+            // Arrange
+            IEdmModel edmModel = EdmModelHelper.GraphBetaModel;
+            ODataContext context = new ODataContext(edmModel);
```
Comment (xuzhg) [inline test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs:482]: It seems "context" is useless?

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs [hunk 2]
```diff
@@ -471,6 +472,44 @@ public void CreatePropertySchemaForNullableEnumPropertyReturnSchema(OpenApiSpecV
...
+            Assert.NotNull(schema.OneOf);
+            Assert.Equal(derivedTypesCount, schema.OneOf.Count);
+        }
+
+        [Fact]
+        public void GetDerivedTypesReferenceSchemaReturnsNullSchemaIfNotExist()
+        {
+            // Arrange
+            IEdmModel model = EdmModelHelper.GraphBetaModel;
+            ODataContext context = new ODataContext(model);
```
Comment (xuzhg) [inline test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs:500]: same here? right?

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs (hunk 1 above)
Comment (irvinesunday) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs]: You're right 👍🏻

File: test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs (hunk 2 above)
Comment (irvinesunday) [reply test/Microsoft.OpenAPI.OData.Reader.Tests/Generator/OpenApiSchemaGeneratorTests.cs]: Yes ✔️

File: src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
+﻿using System.Collections.Generic;
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs:1]: copyright

File: src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+using System.Linq;
+using Microsoft.OData.Edm;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.OpenApi.OData.Common
+{
+    internal static class Helpers
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs:8]: more meaning full class name. for example "xxxxHelpers"

File: src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        /// <summary>
+        /// Adds the derived types references together with their base type reference in the OneOf property of an OpenAPI schema.
+        /// </summary>
+        /// <returns>The OpenAPI schema with the list of derived types references and their base type references set in the OneOf property.</returns>
+        internal static OpenApiSchema GetDerivedTypesReferenceSchema(IEdmEntityType entityType, IEdmModel edmModel)
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs:14]: check the input nullability?

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs [hunk 6]
```diff
@@ -105,28 +105,41 @@ public class OpenApiConvertSettings
         /// </summary>
         public bool EnableDiscriminatorValue { get; set; } = false;
 
+        /// <summary>
+        /// Gets/sets a value indicating whether or not to show the derived types of a reference in the responses payload.
+        /// </summary>
+        public bool ShowDerivedTypesReferencesForResponses { get; set; } = false;
```
Comment (xuzhg) [inline src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs:111]: Maybe use the same pattern for the configuration name?  "Show...." => Enable....

File: src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs (hunk 3 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs]: Added

File: src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs (hunk 4 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs]: Renamed appropriately

File: src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs (hunk 5 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Common/Helpers.cs]: 👍🏻

File: src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs (hunk 6 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/OpenApiConvertSettings.cs]: Effected

**Comment by xuzhg** [comment]
Merged.

**Pull request closed without merging**